### PR TITLE
feat: TLS certificate rotation

### LINKED ISSUE #84: TLS certificate rotation
Implement TLS certificate rotation to provide a tool for clusters created 2+ years ago that are reaching the default cert expiration.

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**Reason for Change**:

Provide a process that for rotating CA, etcd, kubelet, kubeconfig and apiserver certificates in a cluster built with AKS Engine.

**Issue Fixed**:

Fixes #84 

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: docs/topics/certificaterotation.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Cluster certificate rotation
```
Comment (CecileRobertMichon) [inline docs/topics/certificaterotation.md]: doc is still in progress

File: docs/topics/certificaterotation.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+- The etcd members MUST be in healthy state before rotating the CA and certs.
+- The `_output` directory for your cluster deployment, including working SSH key for accessing the master nodes and `kubeconfig`.
+
+## Preparation
+
+**CAUTION**: Before rotation, it's preferrable to back up your cluster:
+
+1. On an etcd node, take a backup of the current state.
+2. //TODO recommend etcd backup tool, extract the existing state from etcd. Copy this back to your working machine as a precaution, in case the control plane goes down.
+3. Keep a copy of your current _output directory
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
3. Keep a copy of your current _output directory.
```

File: docs/topics/certificaterotation.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# Cluster certificate rotation
+
+Instructions on how to rotate TLS CA and certs for an AKS Engine cluster.
+
+## Prerequesites
+
+- Kubernetes version 1.8.0 or later is required
+- The etcd members MUST be in healthy state before rotating the CA and certs.
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
- The etcd members MUST be in a healthy state before rotating the CA and certs.
```

File: docs/topics/certificaterotation.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# Cluster certificate rotation
+
+Instructions on how to rotate TLS CA and certs for an AKS Engine cluster.
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
Instructions on rotating TLS CA and certificates for an AKS Engine cluster.
```

File: docs/topics/certificaterotation.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+## Prerequesites
+
+- Kubernetes version 1.8.0 or later is required
+- The etcd members MUST be in healthy state before rotating the CA and certs.
+- The `_output` directory for your cluster deployment, including working SSH key for accessing the master nodes and `kubeconfig`.
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
- The `_output` directory for your cluster deployment, including a working SSH key for accessing the master nodes and `kubeconfig`.
```

File: docs/topics/certificaterotation.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+## Rotation
+
+**CAUTION**: Rotating certificates will cause cluster downtime
+
+run `aks-engine rotate-certs`
+
+`aks-engine rotate-certs` will:
+
+- Use part of the AKS-Engine deploy code to generate the certificates we need, using the same cert properties and default expiration as new cluster certs
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
- Use a part of the AKS-Engine deploy code to generate the required certificates by using the same certificate properties and default expiration as the new ones.
```

File: docs/topics/certificaterotation.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+- From master rotate etcd CA and certificates in all masters
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
- Rotate etcd CA and certificates in all of the master nodes.
```

File: docs/topics/certificaterotation.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+- From master rotate CA and certs in the nodes
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion

```

Comment (tariq1890) [reply docs/topics/certificaterotation.md]: Unneeded?

File: docs/topics/certificaterotation.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+- Update kubeconfig and rotate the kubelet certs
```
Comment (tariq1890) [inline docs/topics/certificaterotation.md]:
```suggestion
- Update the kubeconfig and rotate the kubelet certificates.
```

File: docs/design/cert-rotation.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+- Requires re-provisionning all the VMs, causing unnecessary time overhead
+- Current implementation does not support this: the newly built master nodes would not be able to join the etcd cluster, thus failing CSE validation
+
+### Bash scripts
+
+What: Provide users with a bash script that they can run to rotate certificates on their cluster.
+
+Pros:
+
+- Easier to implement
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]: Full stops for these phrases please

File: docs/design/cert-rotation.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+What: Implementing a new "certificate-rotation" command that runs upgrade under the hood. This command would perform a same k8s version upgrade on the cluster with newly generated certificates replacing the old ones in the apimodel.
+
+Pros:
+
+- No reboot required after the fact
+
+Cons:
+
+- Requires re-provisionning all the VMs, causing unnecessary time overhead
+- Current implementation does not support this: the newly built master nodes would not be able to join the etcd cluster, thus failing CSE validation
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]:
```suggestion
- Current implementation does not support this: the newly built master nodes would not be able to join the etcd cluster, thus failing CSE validation.
```

File: docs/design/cert-rotation.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+What: Implementing a new "certificate-rotation" command that runs upgrade under the hood. This command would perform a same k8s version upgrade on the cluster with newly generated certificates replacing the old ones in the apimodel.
+
+Pros:
+
+- No reboot required after the fact
+
+Cons:
+
+- Requires re-provisionning all the VMs, causing unnecessary time overhead
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]:
```suggestion
- Requires a re-provisioning of all the VMs, causing an unnecessary time overhead.
```

File: docs/design/cert-rotation.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+## Alternative implementations considered
+
+### Using upgrade to update the certificates
+
+What: Implementing a new "certificate-rotation" command that runs upgrade under the hood. This command would perform a same k8s version upgrade on the cluster with newly generated certificates replacing the old ones in the apimodel.
+
+Pros:
+
+- No reboot required after the fact
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]:
```suggestion
- No reboot required after the fact.
```

File: docs/design/cert-rotation.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+Until PR [#396](https://github.com/Azure/aks-engine/pull/396), the default expiration of the certs generated by AKS Engine was 2 years. Clusters that were deployed with an AKS Engine version prior to that PR will soon reach expiration. We currently do not have an outlined process for rotating certificates in an existing cluster.
+
+## Goals
+
+Provide a process that is easy to follow for rotating CA, etcd, kubelet, kubeconfig and apiserver certificates in a cluster built with AKS Engine.
+
+## Non-Goals
+
+- Providing a certificate rotation tool for AKS clusters
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]: Full stops here, please.

File: docs/design/cert-rotation.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
+# Cluster Certificate Rotation Feature Design
+
+Design doc on the certificate rotation feature for AKS Engine clusters.
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]:
```suggestion
Design document for the certificate rotation feature in AKS Engine clusters.
```

File: docs/design/cert-rotation.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+- Requires more manual steps from the user. For example, they would have to set environemnent variables for their SSH key, KUBECONFIG, _output directory, etc.
+- Not Windows friendly: would require running the script in a container, unless we provide a Powershell equivalent
+
+### New aks-engine binary command
+
+What: New `aks-engine rotate-certs` command that uses a combination of Kubernetes client-go and SSH commands to access the cluster and rotate certificates.
+
+Pros:
+
+- Better UX
```
Comment (tariq1890) [inline docs/design/cert-rotation.md]: I think you'd know what I'm gonna by now :P

File: docs/design/cert-rotation.md (hunk 14 above)
Comment (CecileRobertMichon) [reply docs/design/cert-rotation.md]: you mean periods?

File: docs/design/cert-rotation.md (hunk 16 above)
Comment (CecileRobertMichon) [reply docs/design/cert-rotation.md]: haha

File: docs/design/cert-rotation.md (hunk 14 above)
Comment (CecileRobertMichon) [reply docs/design/cert-rotation.md]: nevermind, it became clear in the following comments :)

Comment (tariq1890) [reply docs/design/cert-rotation.md]: Yeah. I try to refrain from AmE as much as possible :P.

File: docs/topics/certificaterotation.md (hunk 8 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: there's a missing word. I think it should be apiserver certificates.

File: docs/topics/certificaterotation.md (hunk 6 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: I reworded to make the sentence more clear.

File: docs/topics/certificaterotation.md [hunk 17]
```diff
@@ -0,0 +1,52 @@
+# Cluster certificate rotation
+
+Instructions on rotating TLS CA and certificates for an AKS Engine cluster.
+
+## Prerequesites
+
+- Kubernetes version 1.8.0 or later is required
+- The etcd members MUST be in a healthy state before rotating the CA and certs.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: We might want to specify more details. I assume you mean that not only is an elected leader actively serving requests, but that all nodes in the cluster are online and available? Or is it only important that they have a working configuration, and not that they literally have to be online?

File: docs/topics/certificaterotation.md [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+- The `_output` directory for your cluster deployment, including a working SSH key for accessing the master nodes and `kubeconfig`.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: Why do we need the `_output` directory? Is it just for these two requirements? This might be better stated:

- A working ssh private key that has root access to all master nodes, and a working `kubeconfig` file in order to establish an authenticated connection to the Kubernetes cluster API. These files are persisted at AKS Engine template generation time, by default to the `_output/` child directory from the working parent directory at the time of the `aks-engine` invocation.

(or something like that)

File: docs/topics/certificaterotation.md [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+## Preparation
+
+**CAUTION**: Before rotation, it's preferrable to back up your cluster:
+
+1. On an etcd node, take a backup of the current state.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: Do you need to do this on an etcd node literally? Can you do this from anywhere that has an authenticated connection to the etcd cluster?

File: docs/topics/certificaterotation.md (hunk 17 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: I think the easiest way to meet that requirement is to check that when you run `etcdctl cluster-health` all peers are healthy and cluster is healthy.

File: docs/topics/certificaterotation.md (hunk 19 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: I don't know, the etcd backup is mostly TODO, it's not part of the cert-rotation functionality I'm adding, it's a pre-req for which we will provide a tool recommendation (@palma21 was going to look into that). These are just standard instructions I got from the coreos website.

File: docs/topics/certificaterotation.md (hunk 18 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: 👍

File: docs/topics/certificaterotation.md [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+2. //TODO recommend etcd backup tool, extract the existing state from etcd. Copy this back to your working machine as a precaution, in case the control plane goes down.
+3. Keep a copy of your current _output directory.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: Not sure what the distinction is between this guidance and the prerequisite guidance above.

File: docs/topics/certificaterotation.md [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+
+## Rotation
+
+**CAUTION**: Rotating certificates will cause cluster downtime.
+
+run `aks-engine rotate-certs`
+
+`aks-engine rotate-certs` will:
+
+- Use the exisiting certificate generation code to generate new certificates, using the same certificate properties and default expiration as new certificates generated for new clusters.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: Do we need to mention "Use the existing certificate genaration code" here? I think "Generate new certificates" is good enough.

File: docs/topics/certificaterotation.md [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+This can be done by rebooting all the nodes in the cluster.
+
+// TODO: add `aks-engine` reboot command
+
+## Verification
+
+After the above steps, you can verify the success of the CA and certs rotation:
+
+- Old  `kubeconfig`  should  **NOT**  be able to contact the API server, however the new kubeconfig should be able to talk to it.
```
Comment (jackfrancis) [inline docs/topics/certificaterotation.md]: We might consider writing a general purpose script to do this validation (might be useful apart from cert rotation context):

`aks-engine validate --kubeconfig=<path-to-kube-config> --api-model=<path-to-api-model>`

Or maybe we can parse the template/params to get the validation input (number of nodes, etc), either via a local file or via running an `az`-like command against the resource group.

File: docs/topics/certificaterotation.md (hunk 19 above)
Comment (palma21) [reply docs/topics/certificaterotation.md]: Yeah, the user can do it without directly hopping into the node, but that is one option.

Suggest we generalize it as
Backup the etcd cluster data

File: docs/topics/certificaterotation.md (hunk 22 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: Good idea, that would be phase 2 though. Phase 1 will focus on the cert rotation itself.

File: docs/topics/certificaterotation.md (hunk 21 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: Yeah I'll remove that, that was mostly when I was trying to make this a design doc by describing the implementation.

File: docs/topics/certificaterotation.md (hunk 20 above)
Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: true, there isn't one. Will remove one of them.

**Comment by tariq1890** [comment]
What about the certificate rotation provided by K8s. Shouldn't we be using the first-class solutions to this ?

https://kubernetes.io/docs/tasks/tls/certificate-rotation/
https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/

We should be using kubernetes csr's no?

**Comment by CecileRobertMichon** [comment]
@tariq1890 as far as I understand the k8s cert rotation is only for kubelet certificates, whereas we want to rotate all certificates generated by aks-engine deploy (etcd, kubeconfig, etc.). Also the feature is beta in 1.8  (https://kubernetes.io/docs/tasks/tls/certificate-rotation/#before-you-begin) and not available for older k8s versions. Since our goal here is to provide a tool to rotate certificates for old aks-engine clusters (that were created 2 years ago), this won't cover all of them since there might be clusters created 2 years ago that are still on 1.6 or 1.7.

I think we should still provide a tool for old acs/aks engine clusters to rotate their certificates when these certificates expire after the 2 year default expiration (now changed to 30 years in newer clusters). This is what this design doc aims to solve. As a separate feature, we should look into enabling the kubelet certificate rotation feature `-feature-gates=RotateKubeletClientCertificate=true` in new clusters so they can automatically rotate their kubelet cert.

File: docs/topics/certificaterotation.md (hunk 22 above)
Comment (palma21) [reply docs/topics/certificaterotation.md]: @jackfrancis what would that validate command do exactly we can spec it for phase two or independently if it makes more sense.

**Comment by palma21** [comment]
@CecileRobertMichon good to go?

**Comment by CecileRobertMichon** [comment]
@palma21 nope, I'm using the same branch for implementation now that I got feedback on the doc. We don't want to merge the feature doc before having the feature implemented. If we want to merge the design doc first, I can make a separate pr or add a disclaimer to the instructions saying the feature doesn't exist yet.

**Comment by palma21** [comment]
Yes I meant to move fwd with implementation, or if you needed additional feedback. 
Same PR is fine by me, I'd prefer not to have the docs before the implementation is merged (in this or other pr)
Thanks!

File: cmd/rotate_certs.go [hunk 23]
```diff
@@ -0,0 +1,131 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package cmd
```
Comment (CecileRobertMichon) [inline cmd/rotate_certs.go]: Work in progress - not ready for review

**Comment by CecileRobertMichon** [comment]
Example output: 

```
INFO[0004] Generating new certificates
INFO[0015] Rotating apiserver certificate
INFO[0030] Rotating kubelet certificate
INFO[0046] Rotating etcd certificates
INFO[0081] Rebooting all nodes... This might take a few minutes
INFO[0483] Updating kubeconfig
INFO[0495] Successfully rotated etcd and cluster certificates.
```

File: cmd/rotate_certs.go [hunk 24]
```diff
@@ -0,0 +1,500 @@
...
+	if err != nil {
+		return errors.Wrap(err, "failed to get Kubernetes Client")
+	}
+	saList, err := kubeClient.ListServiceAccounts(kubeSystemNamespace)
+	if err != nil {
+		return errors.Wrap(err, "failed to get cluster service accounts in namespace "+kubeSystemNamespace)
+	}
+	for _, sa := range saList.Items {
+		switch sa.Name {
+		case "kube-dns", "kubernetes-dashboard", "metrics-server":
```
Comment (CecileRobertMichon) [inline cmd/rotate_certs.go]: verified that heapster doesn't need to be deleted

Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: TODO: check core-dns

File: pkg/engine/template_generator.go [hunk 25]
```diff
@@ -463,7 +463,7 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) templat
 			if !cs.Properties.OrchestratorProfile.IsKubernetes() {
 				return false
 			}
-			return to.Bool(cs.Properties.OrchestratorProfile.KubernetesConfig.PrivateCluster.Enabled)
+			return cs.Properties.OrchestratorProfile.KubernetesConfig != nil && cs.Properties.OrchestratorProfile.KubernetesConfig.PrivateCluster != nil && to.Bool(cs.Properties.OrchestratorProfile.KubernetesConfig.PrivateCluster.Enabled)
```
Comment (CecileRobertMichon) [inline pkg/engine/template_generator.go]: will revert the private cluster changes once #998 merges

**Comment by jackfrancis** [comment]
lgtm

File: pkg/engine/template_generator.go (hunk 25 above)
Comment (CecileRobertMichon) [reply pkg/engine/template_generator.go]: done

File: cmd/rotate_certs.go [hunk 26]
```diff
@@ -0,0 +1,508 @@
...
+		Long:  rotateCertsLongDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return rcc.run(cmd, args)
+		},
+	}
+
+	f := command.Flags()
+	f.StringVarP(&rcc.location, "location", "l", "", "location the cluster is deployed in (required)")
+	f.StringVarP(&rcc.resourceGroupName, "resource-group", "g", "", "the resource group where the cluster is deployed (required)")
+	f.StringVarP(&rcc.apiModelPath, "apimodel", "", "", "the filepath of the cluster's apimodel generated by aks-engine (required)")
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: I was trying to note the difference between deploy's apimodel, which is the cluster config provided by the user, and this apimodel, which is the apimodel in _output/ with fully expanded definition and all defaults set. What do you think I could do to clarify that?

[the remark this answers is not in the document]
Comment (mboersma) [reply cmd/rotate_certs.go]: Also (and more importantly), `aks-engine upgrade` uses the same syntax as `deploy`, and it is referring to the _output/apimodel.json.

So maybe make the flag `--api-model`, add the `-m` shortcut, and make the description of both `rotate-certs` and `update` to be `path to the generated apimodel file`?

File: cmd/rotate_certs.go [hunk 27]
```diff
@@ -0,0 +1,508 @@
...
+	rcc := rotateCertsCmd{
+		authProvider:       &authArgs{},
+		sshCommandExecuter: executeCmd,
+	}
+
+	command := &cobra.Command{
+		Use:   rotateCertsName,
+		Short: rotateCertsShortDescription,
+		Long:  rotateCertsLongDescription,
+		RunE: func(cmd *cobra.Command, args []string) error {
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: Yes. Copy-paste, thanks for catching!

[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: fixed

File: cmd/rotate_certs.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: Done

**Comment by CecileRobertMichon** [comment]
/hold

until I validate the coredns scenario

**Comment by CecileRobertMichon** [comment]
/hold cancel

validated coredns works

File: cmd/rotate_certs.go (hunk 24 above)
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: done

**Comment by jackfrancis** [comment]
lgtm

File: cmd/root_test.go [hunk 28]
```diff
@@ -9,18 +9,36 @@ import (
...
+	"github.com/Azure/aks-engine/pkg/armhelpers"
 	"github.com/Azure/go-autorest/autorest/azure"
 	"github.com/gofrs/uuid"
 	"github.com/spf13/cobra"
 	ini "gopkg.in/ini.v1"
 )
 
+//mockAuthProvider implements AuthProvider and allows in particular to stub out getClient()
+type mockAuthProvider struct {
+	getClientMock armhelpers.AKSEngineClient
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/root_test.go]: I just moved this from deploy_test.go without changing anything

[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/root_test.go]: I can rename it if you want (I don't mind either way) but it will require more changes to change all its references.

[the remark this answers is not in the document]
Comment (mboersma) [reply cmd/root_test.go]: No, please don't bother changing it: it's just test code so I'm not as concerned with style.

File: cmd/rotate_certs.go [hunk 29]
```diff
@@ -0,0 +1,506 @@
...
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	rotateCertsName             = "rotate-certs"
+	rotateCertsShortDescription = "Rotate certificates on an existing Kubernetes cluster"
+	rotateCertsLongDescription  = "Rotate certificates on an existing Kubernetes cluster"
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: Done

File: cmd/rotate_certs.go [hunk 30]
```diff
@@ -0,0 +1,506 @@
...
+	if _, err = os.Stat(rcc.sshFilepath); os.IsNotExist(err) {
+		return errors.Errorf("specified ssh filepath does not exist (%s)", rcc.sshFilepath)
+	}
+	rcc.setSSHConfig()
+
+	log.Infoln("Rotating apiserver certificate")
+
+	err = rcc.rotateApiserver()
+	if err != nil {
+		return errors.Wrap(err, "error rotating apiserver")
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: Fixed, thanks for pointing that out.

File: cmd/rotate_certs.go [hunk 31]
```diff
@@ -0,0 +1,506 @@
...
+		Use:   rotateCertsName,
+		Short: rotateCertsShortDescription,
+		Long:  rotateCertsLongDescription,
+		RunE:  rcc.run,
+	}
+
+	f := command.Flags()
+	f.StringVarP(&rcc.location, "location", "l", "", "location the cluster is deployed in (required)")
+	f.StringVarP(&rcc.resourceGroupName, "resource-group", "g", "", "the resource group where the cluster is deployed (required)")
+	f.StringVarP(&rcc.apiModelPath, "apimodel", "m", "", "path to the generated apimodel file (required)")
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/rotate_certs.go]: Fixed.

File: cmd/rotate_certs.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (mboersma) [reply cmd/rotate_certs.go]: :100:

File: docs/topics/certificaterotation.md [hunk 32]
```diff
@@ -0,0 +1,57 @@
...
+**CAUTION**: Rotating certificates can break component connectivity and leave the cluster in an unrecoverable state. Before performing any of these instructions on a live cluster, it is preferrable to backup your cluster state and migrate critical workloads to another cluster.
+
+## Rotation
+
+**CAUTION**: Rotating certificates will cause cluster downtime.
+
+run `aks-engine rotate-certs`. For example:
+
+```bash
+CLUSTER="<CLUSTER_DNS_PREFIX>" && bin/aks-engine rotate-certs --apimodel _output/${CLUSTER}/apimodel.json --client-id "<YOUR_CLIENT_ID>" --client-secret "<YOUR_CLIENT_SECRET>" --location <CLUSTER_LOCATION> --master-FQDN ${CLUSTER}.<CLUSTER_LOCATION>.cloudapp.azure.com --ssh _output/${CLUSTER}-ssh --subscription-id "<YOUR_SUBSCRIPTION_ID>" -g ${CLUSTER} -o _output/${CLUSTER}
```
Comment (mboersma) [inline docs/topics/certificaterotation.md]: This command needs to change to `api-model` with a hyphen now. My fault.

Comment (CecileRobertMichon) [reply docs/topics/certificaterotation.md]: fixed

Comment (mboersma) [reply docs/topics/certificaterotation.md]: Also, could you break this up with backslashes into several shorter lines? Currently it renders into markdown as one long line with a horizontal scrollbar.

File: docs/topics/certificaterotation.md [hunk 33]
```diff
@@ -0,0 +1,57 @@
...
+- Reboot all the VMs in the resource group.
+- Restart all the pods to ensure they refresh their service account.
+
+## Verification
+
+After the above steps, you can verify the success of the CA and certs rotation:
+
+- Old  `kubeconfig`  should  **NOT**  be able to contact the API server, however the new kubeconfig should be able to talk to it.
+- All nodes are expected to be  `Ready`, all pods are expected to be  `Running`.
+- Try to fetch the logs of  `kube-apiserver`,  `kube-scheduler`  and  `kube-controller-namager`, they should all be running correctly without spitting errors. E.g. ```kubectl logs -lk8s-app=kube-scheduler -n kube-system`.
```
Comment (mboersma) [inline docs/topics/certificaterotation.md]: s/spitting/printing/

Comment (mboersma) [reply docs/topics/certificaterotation.md]: Could you put a period after kube-controller-manager and start a new sentence?
``` 
... and `kube-controller-manager`. They should all be...
```

Comment (mboersma) [reply docs/topics/certificaterotation.md]: The command here should start with a single backtick to render correctly.

File: docs/topics/certificaterotation.md [hunk 34]
```diff
@@ -0,0 +1,57 @@
...
+The certificate rotation tool has not been tested on and is expected to fail with the following cluster configurations:
+
+- Private clusters.
+- Clusters using keyvault references in Certificate Profile.
+- Clusters using Cosmos etcd.
+- Clusters with already expired certificates with unhealthy etcd.
+
+The rotation involves rebooting the nodes. ALL VMs in the resource group will be restarted as part of running the `rotate-certs` command. If the resource group contains any VMs that are not part of the cluster, they will be restarted as well.
+
+The tool is not currently idempotent, meaning that if the rotation fails halfway though or is interrupted, you will most likely not be able to re-run the operation without any manual intervention. There is a risk that your cluster will become unrecoverable which is why it is strongly recommended to follow the steps in the [preparation step](#preparation).
```
Comment (mboersma) [inline docs/topics/certificaterotation.md]: Remove "any": `re-run the operation without manual intervention.`

We say "step" twice, so maybe reword the last part as:
```
...to follow the [preparation steps](#preparation).
```

File: docs/topics/certificaterotation.md [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+- Rotate etcd CA and certificates and restart etcd in all of the master nodes.
+- Update the kubeconfig.
+- Reboot all the VMs in the resource group.
+- Restart all the pods to ensure they refresh their service account.
+
+## Verification
+
+After the above steps, you can verify the success of the CA and certs rotation:
+
+- Old  `kubeconfig`  should  **NOT**  be able to contact the API server, however the new kubeconfig should be able to talk to it.
```
Comment (mboersma) [inline docs/topics/certificaterotation.md]: I would start with "The": ``` The old `kubeconfig` should ```
